circuit-construction/src/writer.rs

```diff
@@ -122,50 +126,62 @@ pub trait Cs<F: PrimeField> {
     fn assert_eq(&mut self, x1: Var<F>, x2: Var<F>) {
         // | 0  | 1  | 2 | ...
         // | x1 | x2 | 0 | ...
-        let row = vec![Some(x1), Some(x2)];
+        let vars = [Some(x1), Some(x2), None];
 
         // constrain `x1 - x2 = 0`
-        let mut coeffs = vec![F::zero(); DOUBLE_GENERIC_COEFFS];
+        let mut coeffs = [F::zero(); GENERIC_COEFFS];
         coeffs[0] = F::one();
         coeffs[1] = -F::one();
 
-        self.gate(GateSpec {
-            typ: GateType::Generic,
-            row,
-            coeffs,
-        });
+        self.generic(coeffs, vars);
     }
 
     fn constant(&mut self, x: F) -> Var<F> {
         let v = self.var(|| x);
 
-        let mut coeffs = vec![F::zero(); DOUBLE_GENERIC_COEFFS];
+        let mut coeffs = [F::zero(); GENERIC_COEFFS];
         coeffs[0] = F::one();
         coeffs[GENERIC_REGISTERS + 1] = -x;
 
-        let row = vec![Some(v)];
+        let vars = [Some(v), None, None];
+
+        self.generic(coeffs, vars);
 
-        self.gate(GateSpec {
-            typ: GateType::Generic,
-            row,
-            coeffs,
-        });
         v
     }
 
+    /// Stores a generic gate until it can combine two of them
+    /// into a double generic gate.
+    fn generic_queue(&mut self, gate: GateSpec<F>) -> Option<GateSpec<F>>;
+
+    /// Adds a generic gate.
+    ///
+    /// Warning: this assumes that some finalization occurs to flush
+    /// any queued generic gate.
+    fn generic(&mut self, coeffs: [F; GENERIC_COEFFS], vars: [Option<Var<F>>; GENERIC_REGISTERS]) {
+        let gate = GateSpec {
+            typ: GateType::Generic,
+            row: vars.to_vec(),
+            coeffs: coeffs.to_vec(),
+        };
+        // we queue the single generic gate until we have two of them
+        if let Some(double_generic_gate) = self.generic_queue(gate) {
+            self.gate(double_generic_gate);
+        }
+    }
+
     // TODO
     fn scale(&mut self, x: F, v: Var<F>) -> Var<F> {
         let xv = self.var(|| v.val() * x);
-        let row = vec![Some(v), Some(xv)];
 
-        let mut coeffs = vec![F::zero(); DOUBLE_GENERIC_COEFFS];
+        let vars = [Some(v), Some(xv), None];
```

Comment:
If it was a vector already you could leave the `vec![Some(v), Some(xv)]` as before